staking-cli/src/registration.rs

```diff
@@ -34,60 +43,105 @@ pub async fn register_validator(
     bls_key_pair: BLSKeyPair,
-    schnorr_vk: StateVerKey,
+    schnorr_key_pair: StateKeyPair,
 ) -> Result<TransactionReceipt> {
-    let stake_table = StakeTable::new(stake_table_addr, &provider);
-    let (bls_vk_sol, sig_sol) = prepare_bls_payload(&bls_key_pair, validator_address);
-    let schnorr_vk_sol: EdOnBN254PointSol = schnorr_vk.to_affine().into();
-    Ok(stake_table
-        .registerValidator(
-            bls_vk_sol,
-            schnorr_vk_sol,
-            sig_sol.into(),
-            commission.to_evm(),
-        )
-        .send()
-        .await
-        .maybe_decode_revert::<StakeTableErrors>()?
-        .get_receipt()
-        .await?)
+    let stake_table = StakeTableV2::new(stake_table_addr, &provider);
+    let (bls_vk, bls_sig) = prepare_bls_payload(&bls_key_pair, validator_address);
+    let (schnorr_vk, schnorr_sig) = prepare_schnorr_payload(&schnorr_key_pair, validator_address);
+
+    let version = stake_table.getVersion().call().await?.try_into()?;
+    // NOTE: the StakeTableV2 ABI is a superset of the V1 ABI so we can always use the V2 bindings.
+    // There is a race-condition here if the contract is upgraded while this transactions is waiting
+    // to be mined. We're very unlikely to hit this in practice, and since we only perform the upgrade
+    // on decaf this is acceptable.
+    Ok(match version {
+        StakeTableContractVersion::V1 => {
+            stake_table
+                .registerValidator(bls_vk, schnorr_vk, bls_sig.into(), commission.to_evm())
+                .send()
+                .await
+                .maybe_decode_revert::<StakeTableV2Errors>()?
+                .get_receipt()
+                .await?
+        },
+        StakeTableContractVersion::V2 => {
+            stake_table
+                .registerValidatorV2(
+                    bls_vk,
+                    schnorr_vk,
+                    bls_sig.into(),
+                    schnorr_sig,
+                    commission.to_evm(),
+                )
+                .send()
+                .await
+                .maybe_decode_revert::<StakeTableV2Errors>()?
+                .get_receipt()
+                .await?
+        },
+    })
 }
 
 pub async fn update_consensus_keys(
     provider: impl Provider,
     stake_table_addr: Address,
     validator_address: Address,
     bls_key_pair: BLSKeyPair,
-    schnorr_vk: StateVerKey,
+    schnorr_key_pair: StateKeyPair,
 ) -> Result<TransactionReceipt> {
-    let stake_table = StakeTable::new(stake_table_addr, &provider);
-    let (bls_vk_sol, sig_sol) = prepare_bls_payload(&bls_key_pair, validator_address);
-    let schnorr_vk_sol: EdOnBN254PointSol = schnorr_vk.to_affine().into();
-    Ok(stake_table
-        .updateConsensusKeys(bls_vk_sol, schnorr_vk_sol, sig_sol.into())
-        .send()
-        .await
-        .maybe_decode_revert::<StakeTableErrors>()?
-        .get_receipt()
-        .await?)
+    let stake_table = StakeTableV2::new(stake_table_addr, &provider);
```

Comment:
Can we add some comments that since V2 inherits v1, all v1 methods are available and thus the "get_version" layer still make senses, this instantiation won't fail even if the address is a v1 stake table